File: data structures and algorithms/java/08_025.java
import java.util.*;

/**
 * Represents a task with scheduling attributes
 */
class Task {
   String id;
   int priority;
   int resourceNeeds;
   int deadline;
   
   Task(String id, int priority, int resourceNeeds, 
         int deadline) {
      this.id = id;
      this.priority = priority;
      this.resourceNeeds = resourceNeeds;
      this.deadline = deadline;
   }
}

/**
 * Schedules tasks based on priority and deadline
 */
public class AdaptiveResourceScheduler {
   /**
    * Schedule tasks in optimal order
    */
   public static List<String> scheduleTasks(List<Task> tasks) {
      // Sort by priority (high to low) then deadline (early first)
      Collections.sort(tasks, (a, b) -> {
         if (a.priority != b.priority) {
            return b.priority - a.priority; // Descending priority
         }
         return a.deadline - b.deadline; // Ascending deadline
      });
      
      // Store execution order
      List<String> result = new ArrayList<>();
      
      // Track current time
      int currentTime = 0;
      
      // Process each task in sorted order
      for (Task task : tasks) {
         result.add(task.id);
         currentTime += task.resourceNeeds;
      }
      
      return result;
   }
   
   /**
    * Test the scheduling algorithm
    */
   public static void main(String[] args) {
      List<Task> tasks = Arrays.asList(
         new Task("T1", 3, 5, 10),
         new Task("T2", 5, 3, 5),
         new Task("T3", 2, 2, 7),
         new Task("T4", 5, 1, 3)
      );
      
      List<String> schedule = scheduleTasks(tasks);
      System.out.println("Task execution order: " + schedule);
      // Expected output: [T4, T2, T1, T3]
   }
}

File: data structures and algorithms/cs/08_025.cs
using System;
using System.Collections.Generic;

class Task {
    public string Id { get; set; }
    public int Priority { get; set; }
    public int ResourceNeeds { get; set; }
    public int Deadline { get; set; }

    public Task(string id, int priority, int resourceNeeds, int deadline) {
        Id = id;
        Priority = priority;
        ResourceNeeds = resourceNeeds;
        Deadline = deadline;
    }
}

public class AdaptiveResourceScheduler {
    public static List<string> ScheduleTasks(List<Task> tasks) {
        tasks.Sort((a, b) => {
            if (a.Priority != b.Priority) {
                return b.Priority - a.Priority;
            }
            return a.Deadline - b.Deadline;
        });

        var result = new List<string>();
        int currentTime = 0;

        foreach (var task in tasks) {
            result.Add(task.Id);
            currentTime += task.ResourceNeeds;
        }

        return result;
    }

    public static void Main(string[] args) {
        var tasks = new List<Task> {
            new Task("T1", 3, 5, 10),
            new Task("T2", 5, 3, 5),
            new Task("T3", 2, 2, 7),
            new Task("T4", 5, 1, 3)
        };

        var schedule = ScheduleTasks(tasks);
        Console.WriteLine("Task execution order: " + string.Join(", ", schedule));
    }
}